components/script/dom/toggleevent.rs

```diff
@@ -0,0 +1,145 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use dom_struct::dom_struct;
+use js::rust::HandleObject;
+use script_bindings::codegen::GenericBindings::NodeBinding::NodeMethods;
+use script_bindings::inheritance::Castable;
+use stylo_atoms::Atom;
+
+use crate::dom::bindings::codegen::Bindings::EventBinding::EventMethods;
+use crate::dom::bindings::codegen::Bindings::ToggleEventBinding;
+use crate::dom::bindings::codegen::Bindings::ToggleEventBinding::ToggleEventMethods;
+use crate::dom::bindings::error::Fallible;
+use crate::dom::bindings::reflector::reflect_dom_object_with_proto;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::element::Element;
+use crate::dom::event::{Event, EventBubbles, EventCancelable};
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::node::Node;
+use crate::dom::types::Window;
+use crate::script_runtime::CanGc;
+
+#[dom_struct]
+pub(crate) struct ToggleEvent {
+    event: Event,
+    old_state: DOMString,
+    new_state: DOMString,
+    source: Option<DomRoot<Element>>,
+}
+
+#[allow(non_snake_case)]
+impl ToggleEvent {
+    pub(crate) fn new_inherited(
+        old_state: DOMString,
+        new_state: DOMString,
+        source: Option<DomRoot<Element>>,
+    ) -> ToggleEvent {
+        ToggleEvent {
+            event: Event::new_inherited(),
+            old_state,
+            new_state,
+            source,
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        window: &Window,
+        type_: Atom,
+        bubbles: EventBubbles,
+        cancelable: EventCancelable,
+        old_state: DOMString,
+        new_state: DOMString,
+        source: Option<DomRoot<Element>>,
+        can_gc: CanGc,
+    ) -> DomRoot<ToggleEvent> {
+        Self::new_with_proto(
+            window, None, type_, bubbles, cancelable, old_state, new_state, source, can_gc,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn new_with_proto(
+        window: &Window,
+        proto: Option<HandleObject>,
+        type_: Atom,
+        bubbles: EventBubbles,
+        cancelable: EventCancelable,
+        old_state: DOMString,
+        new_state: DOMString,
+        source: Option<DomRoot<Element>>,
+        can_gc: CanGc,
+    ) -> DomRoot<ToggleEvent> {
+        let event = Box::new(ToggleEvent::new_inherited(old_state, new_state, source));
+        let ev = reflect_dom_object_with_proto(event, window, proto, can_gc);
```

Comment:
It's a bit confusing to have three variables, two named `event` and one named `ev` so I would either use `event` for all of them or use a different one non-abbreviated here.

```suggestion
        let event = reflect_dom_object_with_proto(event, window, proto, can_gc);
```